builtin/logical/pki/cert_util.go

```diff
@@ -674,7 +674,8 @@ func generateCert(sc *storageContext,
 		data.Params.PermittedDNSDomains = input.apiData.Get("permitted_dns_domains").([]string)
 
 		if data.SigningBundle == nil {
-			// Generating a self-signed root certificate
+			// Generating a self-signed root certificate. Since we have no
+			// issuer entry yet, we default to the global URLs.
 			entries, err := getURLs(ctx, sc.Storage)
```

Comment:
I guess this approach works for now, but it means there are certain configurations it will be hard to orchestrate reasonably with desired state configuration e.g. terraform-provider-vault. (I'm thinking of ones where you deliberately want to create multiple issuers with different URLs.)

It may be worth re-visiting allowing the URLs to be passed in as part of the root generation request itself in a later PR.

Reply:
I _think_ there should be no issue once `terraform-provider-vault` is updated to support setting AIAs on issuers? (cc @stevendpclark & @benashz).

My understanding from the RFC -- and correct me if I'm wrong -- is generally you don't want AIAs on roots. So while we don't support setting them on roots (as you mention in the last sentence), the approach you'd generally want is:

 1. Generate root without AIA URLs.
 2. Configure per-issuer AIA on the root.
 3. Generate intermediates. These will use the root's per-issuer AIA for signing the intermediate cert proper.
 4. Once intermediate is imported, configure per-issuer AIA on the intermediate.
 5. Issue leaf certs, which will consume intermediate AIAs.

AIA should then be structured correctly, minus the cross-cluster discussion elsewhere. 

---

This comment is mostly about preserving backwards compatibility and noting why we call `getURLs` rather than the newer issuer-entry call (used above).